src/main/java/delight/nashornsandbox/internal/GraalSandboxImpl.java

```diff
@@ -0,0 +1,127 @@
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Predicate;
+
+import javax.script.Bindings;
+import javax.script.ScriptContext;
+import javax.script.ScriptException;
+
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.HostAccess;
+import org.graalvm.polyglot.PolyglotAccess;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.oracle.truffle.js.scriptengine.GraalJSScriptEngine;
+
+import delight.nashornsandbox.exceptions.ScriptCPUAbuseException;
+
+
+/**
+ * Nashorn sandbox implementation for GraalJS
+ * @author marcoellwanger
+ */
+public class GraalSandboxImpl extends NashornSandboxImpl {
+
+	static final Logger LOG = LoggerFactory.getLogger(GraalSandboxImpl.class);
+	
+	public final boolean isStrict;
+	
+	public GraalSandboxImpl() {
+		this(new String[0]);
+	}
+	
+	public GraalSandboxImpl(String... params) {
+		super(GraalJSScriptEngine.create(null, Context.newBuilder().allowExperimentalOptions(true).allowPolyglotAccess(PolyglotAccess.ALL).allowHostAccess(HostAccess.ALL)), params);
+		isStrict = Arrays.asList(params).contains("-strict");
+		Bindings bindings = this.scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+		bindings.put("polyglot.js.allowHostClassLookup", (Predicate<String>) s -> this.sandboxClassFilter.getStringCache().contains(s));
+	}
+	
+	@Override
+	protected void resetEngineBindings() {
+		
+    }
+	
+	@Override
+	public Bindings createBindings() {
+		return scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+	}
+	
+    protected void produceSecureBindings() {
+        try {
+            final StringBuilder sb = new StringBuilder();
+            cached = scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+            sanitizeBindings(cached);
+            if (!allowExitFunctions) {
+                sb.append("var quit=function(){};var exit=function(){};");
+            }
+            if (!allowPrintFunctions) {
+                sb.append("var print=function(){};var echo = function(){};");
+            }
+            if (!allowReadFunctions) {
+                sb.append("var readFully=function(){};").append("var readLine=function(){};");
+            }
+            if (!allowLoadFunctions) {
+                sb.append("var load=function(){};var loadWithNewGlobal=function(){};");
+            }
+            if (!allowGlobalsObjects) {
+                // Max 22nd of Feb 2018: I don't think these are strictly necessary since they are only available in scripting mode
+                sb.append("var $ARG=null;var $ENV=null;var $EXEC=null;");
+                sb.append("var $OPTIONS=null;var $OUT=null;var $ERR=null;var $EXIT=null;");
+            }
+            scriptEngine.eval(sb.toString());
+            this.engineAsserted = true;
+        }
+        catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+    
+    @Override
+	protected Bindings secureBindings(Bindings bindings) {
```

Comment:
Currently we can only use one set of bindings in GraalJS due to the limitations that objects cannot be shared across bindings/contexts (except for primitives). Once Graal adds support for this we can remove these overrides.

Reply:
Just to confirm the behaviour here then,

If a binding for "a" -> "b" is passed in to eval, will that persist between each call of eval? 

If it does that is not really desired since that can lead to a leak from one script to another.

Reply:
You're right - I fixed the bindings leak from one script to another. Thanks!